### PR TITLE
Implement s3 filesystem

### LINKED ISSUE #2077: Copy/Load from remote file system (S3)

### PR DESCRIPTION
**Pull request by acquamarin**
This PR implements the S3 filesystem. Similar to HTTP file system, each read request will be wrapped in a http request and sent to the s3 server.
close #2077

### CODE REVIEW

File: extension/httpfs/test/test_files/http.test [hunk 1]
```diff
@@ -46,24 +46,3 @@ Noura|25
...
--STATEMENT call current_setting('s3_accesS_key_id') return *;
----- 1
-kuzuS3keyid
--STATEMENT CALL s3_access_key_id='anotherkey'
----- ok
--STATEMENT call current_setting('s3_access_Key_id') return *;
----- 1
-anotherkey
-
--CASE DuplicateOption
```
Comment (ray6080) [inline extension/httpfs/test/test_files/http.test:64]: Why remove this test?

File: extension/httpfs/include/httpfs.h
Comment (ray6080) [inline extension/httpfs/include/httpfs.h:1]: Can we put `include` under `src`, so it's more consistent with non-extension code?

File: extension/httpfs/include/httpfs.h [hunk 2]
```diff
@@ -30,7 +45,9 @@ struct HTTPParams {
 struct HTTPFileInfo : public common::FileInfo {
     HTTPFileInfo(std::string path, common::FileSystem* fileSystem, int flags);
 
-    void initializeClient();
+    virtual void initialize();
+
+    virtual void initializeClient();
```
Comment (ray6080) [inline extension/httpfs/include/httpfs.h:50]: Double check if this function needs to be public.

File: extension/httpfs/include/s3fs.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+
+public:
+    std::unique_ptr<common::FileInfo> openFile(const std::string& path, int flags,
+        main::ClientContext* context = nullptr,
+        common::FileLockType lock_type = common::FileLockType::NO_LOCK) override;
+
+    bool canHandleFile(const std::string& path) override;
+
+    static std::string URLEncode(const std::string& input, bool encodeSlash = false);
```
Comment (ray6080) [inline extension/httpfs/include/s3fs.h:51]: URLEncode -> encodeURL.

File: extension/httpfs/include/s3fs.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    static ParsedS3URL s3URLParse(std::string url, S3AuthParams& params);
```
Comment (ray6080) [inline extension/httpfs/include/s3fs.h:53]: s3URLParse -> parseS3URL.

File: extension/httpfs/third_party/mbedtls/include/CMakeLists.txt
Comment (ray6080) [inline extension/httpfs/third_party/mbedtls/include/CMakeLists.txt:1]: As discussed, I don't think we need this cmake file, and also other dummy files.

File: src/common/string_utils.cpp [hunk 5]
```diff
@@ -99,5 +99,31 @@ std::string StringUtils::removeEscapedCharacters(const std::string& input) {
     return resultStr;
 }
 
+// Jenkins hash function: https://en.wikipedia.org/wiki/Jenkins_hash_function
+uint64_t StringUtils::hashOP(const std::string& str) {
```
Comment (ray6080) [inline src/common/string_utils.cpp:103]: hashOP  -> caseInsensitiveHash

File: src/common/string_utils.cpp [hunk 6]
```diff
@@ -99,5 +99,31 @@ std::string StringUtils::removeEscapedCharacters(const std::string& input) {
...
+        hash += hash << 10;
+        hash ^= hash >> 6;
+    }
+    hash += hash << 3;
+    hash ^= hash >> 11;
+    hash += hash << 15;
+    return hash;
+}
+
+bool StringUtils::equalsIgnoreCase(const std::string& left, const std::string& right) {
```
Comment (ray6080) [inline src/common/string_utils.cpp:116]: equalsIgnoreCase -> caseInsensitiveEquals

File: src/common/string_utils.cpp [hunk 7]
```diff
@@ -99,5 +99,31 @@ std::string StringUtils::removeEscapedCharacters(const std::string& input) {
     return resultStr;
 }
 
+// Jenkins hash function: https://en.wikipedia.org/wiki/Jenkins_hash_function
```
Comment (ray6080) [inline src/common/string_utils.cpp:102]: Be more specific here that we transform each character to its lower case and apply Jenkins hash function on converted lower case characters.

File: src/include/common/types/timestamp_t.h [hunk 8]
```diff
@@ -62,55 +62,57 @@ struct timestamp_sec_t : public timestamp_t { // NO LINT
 
 // The Timestamp class is a static class that holds helper functions for the Timestamp type.
 // timestamp/datetime uses 64 bits, high 32 bits for date and low 32 bits for time
-class Timestamp {
+class KUZU_API Timestamp {
```
Comment (ray6080) [inline src/include/common/types/timestamp_t.h:65]: Should we move `KUZU_API` to the class level? I think we have functions not intended as APIs, such as `getTimestampConversionExceptionMsg`.

File: src/main/database.cpp [hunk 9]
```diff
@@ -117,7 +117,7 @@ void Database::openLockFile() {
     } else {
         getLockFileFlagsAndType(systemConfig.readOnly, false, flags, lock);
     }
-    lockFile = vfs->openFile(lockFilePath, flags, lock);
+    lockFile = vfs->openFile(lockFilePath, flags, nullptr, lock);
```
Comment (ray6080) [inline src/main/database.cpp:120]: Add an inline comment for `nullptr`.

**Review by ray6080 — APPROVED** [review]

File: extension/httpfs/include/httpfs.h (hunk 2 above)
Comment (acquamarin) [reply extension/httpfs/include/httpfs.h:50]: I think it is a struct and we never write visibility specifier for struct

**Pull request merged**